BUG : np.sum silently drops keepdims for sub-classes of ndarray

tacaswell opened the pull request:
change test from `type(a) is not mu.ndarray` to `not isinstance(a, mu.ndarray)`

This bug can be demonstrated by:

```
class test_array(ndarray):
    pass

base_ = np.arange(30).reshape(5, 6)
tt = base_.view(test_array)

def sum_test(a):
    print(np.sum(a))
    print(np.sum(a, keepdims=True))

sum_test(base_)
sum_test(tt)
```

tacaswell committed:
BUG: many functions silently drop `keepdims` kwarg

change test from `type(a) is not mu.ndarray` to
`not isinstance(a, mu.ndarray)`

Because every sub-class of ndarray is not guaranteed to implement
`keepdims` as a kwarg, when wrapping these methods care must be taken.

The previous behavior was to silently eat the kwarg when dealing with
a sub-class of ndarray.

Now, if `keepdims=np._NoValue` (the new default) it is not passed
through to the underlying function call (so the default value of
`keepdims` is now controlled by the sub-class).  If `keepdims` is not
`np._NoValue` then it is passed through and will raise an exception if
the sub-class does not support the kwarg.

A special case in nanvar was required to deal with `matrix` that previously
relied on `fromnumeric` silently dropping `keepdims`.

charris commented:
Note that there is going to be a `np._NoValue` class that can be used as a default vulue to arguments where we need to check if the the user has passed a value.

charris commented:
@tacaswell `np._NoValue` is available to replace `_NotGiven`. I think this can go into 1.11, but needs comments in the 1.11 release notes and the commits need a squashing. Anyone else have comments?

@juliantaylor We have to assume that subclasses are not maintained.

tacaswell commented:
Sorry for missing the comment in March.

What sort of comments do you want?

The API break that @juliantaylor is asking about is that if the user has a sub-class that does not deal with a `keepdims` kwarg _and_ they are passing a non-default value of `keepdims` to these functions on top of their sub class, it will now raise where as before it would silently drop the keyword and possibly silently do the wrong thiny?  I strongly argue that if that is the case, then the user code was buggy to begin with.

tacaswell commented:
todo

 - [ ] use `np._noValue`
 - [ ] make sure `np.nanprod` is handled

tacaswell commented:
@charris where do the docs go?

tacaswell commented:
32bit windows is failing on the following:

```
 
======================================================================
FAIL: test_nans_infs (test_half.TestHalf)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Py\lib\site-packages\numpy\core\tests\test_half.py", line 78, in test_nans_infs
    assert_equal(np.signbit(self.all_f16), np.signbit(self.all_f32))
  File "C:\Py\lib\site-packages\numpy\testing\utils.py", line 316, in assert_equal
    return assert_array_equal(actual, desired, err_msg, verbose)
  File "C:\Py\lib\site-packages\numpy\testing\utils.py", line 807, in assert_array_equal
    verbose=verbose, header='Arrays are not equal')
  File "C:\Py\lib\site-packages\numpy\testing\utils.py", line 733, in assert_array_compare
    raise AssertionError(msg)
AssertionError: 
Arrays are not equal
 
(mismatch 0.0030517578125%)
 x: array([False, False, False, ...,  True,  True,  True], dtype=bool)
 y: array([False, False, False, ...,  True, False, False], dtype=bool)
 
----------------------------------------------------------------------

```

I see the same error in https://ci.appveyor.com/project/charris/numpy/build/1.0.422/job/ut2jgg6yndrroxvs#L8176 from https://github.com/numpy/numpy/pull/6938 which makes me think that I did not break this.

charris commented:
Yeah that looks like my mistake in not waiting on the tests :(  It probably comes from @juliantaylor recent commit vectorizing signbit.

jakirkham commented:
That's what I was thinking. There were a few other failures on AppVeyor and I have put them in this issue ( https://github.com/numpy/numpy/issues/6991 ).

tacaswell commented:
Commits squashed and made the special case in `nanvar` made a bit nicer, but I am still not sure exactly where to add the documentation.

seberg commented:
I think the "documentation" was just adding a short mention in the compatibility section of the release notes probably. They are in doc/release/. Did not read the code, but seems a bit of a shame that we stalled this for so long.

mhvk commented on numpy/core/fromnumeric.py:
```diff
@@ -1826,15 +1835,14 @@ def sum(a, axis=None, dtype=None, out=None, keepdims=False):
             sum = a.sum
         except AttributeError:
             return _methods._sum(a, axis=axis, dtype=dtype,
```
Sorry to raise a stylistic point at this late stage, but since the `return _methods._sum` call is now identical to the one at the end, I would suggest rewriting this (as well as other occurrences further down) as:
```
# no need to elif here; always returned output above
if type(a) is not mu.ndarray:
    try:
        sum = a.sum
    except AttributeError:
        pass
    else:
        return sum(axis=axis, dtype=dtype, out=out, **kwargs)

return _methods._sum(a, axis=axis, dtype=dtype, out=out, **kwargs)
```

tacaswell committed:
STY: pep8 only

tacaswell committed:
MNT: move std, var, mean calls out of try block

Move the calls to user-provided versions of std, var, and mean on non
mu.ndarray objects out of the `try` block so that numpy will not mask
AttributeError raised during the execution of the function rather than
due to the object not having the required method.

tacaswell committed:
MNT: reduce number of return statements

In sum, amax, amin, and prod simplify the logic to remove an identical
return statement / call to `_methods._xxx`.  This removes several
elif/else pairs and reduces the number of exit points from the functions
but makes the code path a bit more complicated to trace.

tacaswell committed:
DOC: add release notes

tacaswell commented:
```
======================================================================
FAIL: test_scripts.test_f2py
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/numpy/numpy/venv-for-wheel/lib/python3.5/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/home/travis/build/numpy/numpy/venv-for-wheel/lib/python3.5/site-packages/numpy/testing/decorators.py", line 147, in skipper_func
    return f(*args, **kwargs)
  File "/home/travis/build/numpy/numpy/venv-for-wheel/lib/python3.5/site-packages/numpy/tests/test_scripts.py", line 93, in test_f2py
    assert_(success, msg)
  File "/home/travis/build/numpy/numpy/venv-for-wheel/lib/python3.5/site-packages/numpy/testing/utils.py", line 71, in assert_
    raise AssertionError(smsg)
AssertionError: Warning: neither f2py nor f2py3 nor f2py3.5 found in path
----------------------------------------------------------------------
```

Two tests are failing on the above which I suspect is un-related to these changes.

seberg commented:
@charris do you want to backport this? While I like it, I think the default is probably no backport, which would mean that it should be in the 1.12 release notes.

Don't worry about those test failures, some setuptools stuff I don't even want to know about :).

njsmith commented:
Yeah, let's leave this for 1.12 -- it *probably* won't create any surprises
downstream, but it's the sort of thing that could.

On Fri, Feb 5, 2016 at 11:52 AM, seberg <notifications@github.com> wrote:

> @charris <https://github.com/charris> do you want to backport this? While
> I like it, I think the default is probably no backport, which would mean
> that it should be in the 1.12 release notes.
>
> Don't worry about those test failures, some setuptools stuff I don't even
> want to know about :).
> EDIT: or maybe it was venv I guess, you see my point of not wanting to
> know ;)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numpy/numpy/pull/4619#issuecomment-180526557>.
>



-- 
Nathaniel J. Smith -- https://vorpus.org <http://vorpus.org>

tacaswell commented:
done

homu commented:
:umbrella: The latest upstream changes (presumably #7181) made this pull request unmergeable. Please resolve the merge conflicts.

tacaswell commented on numpy/lib/nanfunctions.py:
```diff
@@ -163,9 +164,14 @@ def nanmin(a, axis=None, out=None, keepdims=False):
 
         .. versionadded:: 1.8.0
     keepdims : bool, optional
-        If this is set to True, the axes which are reduced are left in the
```
this is due to line wrap changes in #7181

charris commented:
Backported for 1.11.1.
